### PR TITLE
sparse vector views remain sparse

### LINKED ISSUE #402: Subarrays of sparse should operate like sparse
```julia
M = spzeros(4,4)

julia>  @view(M[:,i]) + @view(M[:,k])
4-element Vector{Float64}:
 0.0
 0.0
 0.0
 0.0

julia>  (M[:,i]) + (M[:,k])
4-element SparseVector{Float64, Int64} with 0 stored entries

```

### PR DESCRIPTION
**Pull request by matbesancon**
Closes #402 for sparse vectors. Closes #225. Fixes https://github.com/JuliaLang/julia/issues/46355.

I also relaxed the assumption that the view indexing must be a Base.OneTo, otherwise this is too restrictive for vectors

### CODE REVIEW

**Comment by matbesancon** [comment]
Alright this is good for review on my side

**Comment by matbesancon** [comment]
@ViralBShah this could be a step towards fixing the views: https://github.com/JuliaSparse/SparseArrays.jl/pull/225#issuecomment-1218070104

File: src/sparsevector.jl [hunk 1]
```diff
@@ -1558,6 +1596,22 @@ for (fun, mode) in [(:+, 1), (:-, 1), (:*, 0), (:min, 2), (:max, 2)]
     end
 end
 
+for fun in (:+, :-)
+    @eval @propagate_inbounds function $(fun)(x::Union{SparseVectorUnion{Tx},SparseVectorPartialView{Tx}}, y::Union{SparseVectorUnion{Ty},SparseVectorPartialView{Ty}}) where {Tx, Ty}
+        @boundscheck size(x) == size(y)
```
Comment (fredrikekre) [inline src/sparsevector.jl:1601]: This doesn't throw on mismatching size, and I think what you should check is `axes`.

Comment (fredrikekre) [reply src/sparsevector.jl]: Still doesn't throw, you need something like `axes(x) == axes(y) || throw(...)`.

**Comment by matbesancon** [comment]
still haven't figured out what is causing the issue here

File: src/sparsevector.jl [hunk 2]
```diff
@@ -114,6 +117,9 @@ function nnz(x::SparseColumnView)
     return length(nzrange(parent(x), colidx))
 end
 nnz(x::SparseVectorView) = nnz(x.parent)
+function nnz(x::SparseVectorPartialView)
```
Comment (SobhanMP) [inline src/sparsevector.jl:120]: some bikeshedding :P but write this as a one liner IMHO

File: src/sparsevector.jl [hunk 3]
```diff
@@ -135,6 +141,12 @@ function nonzeros(x::SparseColumnView)
 end
 nonzeros(x::SparseVectorView) = nonzeros(parent(x))
 
+function nonzeros(x::SparseVectorPartialView)
+    (first_idx, last_idx) = _partialview_end_indices(x)
+    nzvals = nonzeros(parent(x))
+    return @view(nzvals[first_idx:last_idx])
```
Comment (SobhanMP) [inline src/sparsevector.jl:147]: I vaguely remember people saying that you should use the view function directly.

File: src/sparsevector.jl [hunk 4]
```diff
@@ -145,10 +157,36 @@ function nonzeroinds(x::SparseColumnView)
 end
 nonzeroinds(x::SparseVectorView) = nonzeroinds(parent(x))
 
+# return the first and last nonzero indices of the parent that belong to the view
+# return end+1:end if no nonzero in the parent
+function _partialview_end_indices(x::SparseVectorPartialView)
+    p = parent(x)
+    nzinds = nonzeroinds(p)
+    if isempty(nzinds)
+        first_idx = nzinds[begin]
```
Comment (SobhanMP) [inline src/sparsevector.jl:166]: I'm confused, shouldn't this be out of bounds? It's empty.

**Comment by SobhanMP** [comment]
Could you also update the

File: src/sparsevector.jl (hunk 4 above)
Comment (matbesancon) [reply src/sparsevector.jl]: good catch, I added a test to cover the case too

File: src/sparsevector.jl (hunk 3 above)
Comment (matbesancon) [reply src/sparsevector.jl]: done, might be linked to compile time?

File: src/sparsevector.jl (hunk 2 above)
Comment (matbesancon) [reply src/sparsevector.jl]: done indeed

**Comment by dkarrasch** [comment]
I copied over the tests from #225 and thereby found a little oversight. I'll push the easy fix right after, and then we can merge after tests pass.

**Pull request merged**